### PR TITLE
[camera_web] Recording Video

### PR DESCRIPTION
**Pull request by ABausG**
Adds ability to record `webm` of the camera platform interface.

Implements calls `startVideoRecording`, `pauseVideoRecording`, `resumeVideoRecording` and `stopVideoRecording`

Part of flutter/flutter#45297.

## Pre-launch Checklist

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x] I read and followed the [relevant style guides] and ran [the auto-formatter]. (Note that unlike the flutter/flutter repo, the flutter/plugins repo does use `dart format`.)
- [x] I signed the [CLA].
- [x] The title of the PR starts with the name of the plugin surrounded by square brackets, e.g. `[shared_preferences]`
- [x] I listed at least one issue that this PR fixes in the description above.
- [ ] I [updated pubspec.yaml](https://github.com/flutter/flutter/wiki/Contributing-to-Plugins-and-Packages#version-and-changelog-updates) with an appropriate new version according to the [pub versioning philosophy].
- [ ] I updated CHANGELOG.md to add a description of the change.
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] I added new tests to check the change I am making or feature I am adding, or Hixie said the PR is test exempt.
- [x] All existing and new tests are passing.

### CODE REVIEW

**Comment by ABausG** [comment]
@bselwe @felangel @ditman I saw that you did some work on camera_web recently. Could you take a look at this?

File: packages/camera/camera_web/lib/src/camera.dart
Comment (ditman) [inline packages/camera/camera_web/lib/src/camera.dart:238]: There seems to be some code duplication here vs startVideoRecording (with a max length).

Could you unify these two codepaths so the implementation of the`dataavailable` event on `startVideoRecording` uses the same logic as "stop"?

I also wonder why the start with max length emits a `VideoRecordedEvent`, but `stopVideoRecording` doesn't?

File: packages/camera/camera_web/lib/src/camera.dart
Comment (ditman) [inline packages/camera/camera_web/lib/src/camera.dart:203]: What happens if `maxVideoDuration` is `0 milliseconds`? Should that be disallowed?

File: packages/camera/camera_web/lib/src/camera.dart
Comment (ditman) [inline packages/camera/camera_web/lib/src/camera.dart:223]: (These exceptions are already thrown by the browser, no need to assert for _mediaRecorder.state)

**Review by ditman — COMMENTED** [review]
Thanks @ABausG for the contribution, this is great!

I've left some questions about the current approach (mostly about cross-browser compat).

BTW, @bselwe is currently OOTO (until the end of the week), but he was working in normalizing the exceptions thrown from the code.

In this PR, for example, we're manually throwing DomExceptions from the low-level Camera object, but I think @bselwe was working on getting some standardization in *what* types of exceptions to throw *from where*. The guiding lines here would be the core `camera` plugin, seeing what types of Exceptions are expected there, so everything works transparently.

I'll defer to @bselwe for more comments in that regard!

Other than that, this is looking amazing! Thanks again for the contribution!

**Review by bselwe — COMMENTED** [review]
Awesome, thanks for the contribution @ABausG and your comments @ditman! 🙌

I've added some clarification regarding which exceptions to throw and when. The general rule would be to try to match the exceptions thrown in [`MethodChannelCamera`](https://github.com/flutter/plugins/blob/08c9b40d020bf249f346322ec395220e36889b42/packages/camera/camera_platform_interface/lib/src/method_channel/method_channel_camera.dart#L23) (an implementation of the camera platform used for iOS and Android platforms) and handled in [`CameraController`](https://github.com/flutter/plugins/blob/08c9b40d020bf249f346322ec395220e36889b42/packages/camera/camera/lib/src/camera_controller.dart#L204). Let me know if there's anything unclear about this approach.

I think it would also be nice to include `Camera` tests. That could include:
- For `startVideoRecording`:
  - Throws an exception when the media recorder is in an invalid state.
  - Initializes a media recorder instance from the video element stream.
  - Uses an appropriate mime type for the video recording based on `MediaRecorder.isTypeSupported()`.
  - Uses `maxVideoDuration` to record a video.
  - Emits a `VideoRecordedEvent` when the video recording has been stopped.
- For `pauseVideoRecording`:
  - Throws an exception when the media recorder is null.
  - Calls pause on the media recorder.
- And similar tests for other methods.

Thanks again!

**Review by bselwe — COMMENTED** [review]
Should we maybe remove throwing an unimplemented error in [`prepareForVideoRecording`](https://github.com/flutter/plugins/pull/4210/files#diff-95b0d214cebd76a613f66fbe94ad5a3cbccd21e0084d34b83b8589d2a14b5ed3R355)?

**Comment by ABausG** [comment]
> Should we maybe remove throwing an unimplemented error in [`prepareForVideoRecording`]

I was thinking about that method as well. I'm seeing two options:
1. Make it no-op
2. Perform initialisation in the actual camera implementation like setting up the media recorder.

File: packages/camera/camera_web/lib/src/camera.dart
Comment (ABausG) [reply packages/camera/camera_web/lib/src/camera.dart]: I've added a List of possible mimeTypes and changed to use the first supported one. Right now I've added `video/mp4` and `video/webm` not sure what formats/order to add them. Any advice/suggestion?

Comment (ABausG) [reply packages/camera/camera_web/lib/src/camera.dart]: Also using the `srcObject` as suggested by @bselwe

File: packages/camera/camera_web/lib/src/camera.dart
Comment (ABausG) [reply packages/camera/camera_web/lib/src/camera.dart]: Throwing a `PlatformException` now and not checking fo the `state` of the mediaRecorder`

File: packages/camera/camera_web/lib/src/camera.dart
Comment (ABausG) [reply packages/camera/camera_web/lib/src/camera.dart]: Added mimeType and Filename. For now using the `hashcode` of the blob as filename

File: packages/camera/camera_web/lib/src/camera.dart
Comment (ABausG) [reply packages/camera/camera_web/lib/src/camera.dart]: See above. Now throwing PlatformExceptions

File: packages/camera/camera_web/lib/src/camera.dart
Comment (ABausG) [reply packages/camera/camera_web/lib/src/camera.dart]: Unified it by registering the listener on `dataavailable` only in the startVideoRecording.

If the listener is triggered it will do the following things:
- emit a `VideoRecordedEvent`
- Complete the Completer used to obtain the `XFile` in `stopVideoRecording`
- remove the listener
- stop the MediaRecorder
   - This stop is necessary to only get data once if a `maxVideoDuration` is provided as `MediaRecorder` only takes in splices instead of a maxDuration

File: packages/camera/camera_web/lib/src/camera.dart
Comment (ABausG) [reply packages/camera/camera_web/lib/src/camera.dart]: Throwing a PlatformException

File: packages/camera/camera_web/lib/src/camera.dart
Comment (ABausG) [reply packages/camera/camera_web/lib/src/camera.dart]: See above

File: packages/camera/camera_web/lib/src/camera.dart
Comment (ABausG) [reply packages/camera/camera_web/lib/src/camera.dart]: Yes. Removing it now

File: packages/camera/camera_web/lib/src/camera.dart
Comment (ABausG) [reply packages/camera/camera_web/lib/src/camera.dart]: I renamed it to `onVideoRecordedEvent` to match the naming of the getter in the `camera_plugin_interface`

File: packages/camera/camera_web/lib/src/camera.dart
Comment (ditman) [reply packages/camera/camera_web/lib/src/camera.dart]: /cc @bselwe look, YUV 4:2:0 and friends are starting to show up here: https://developer.mozilla.org/en-US/docs/Web/Media/Formats/codecs_parameter#av1 (I don't think we should mess with the `codecs` value, unless strictly necessary)

File: packages/camera/camera_web/example/integration_test/camera_test.dart
Comment (ditman) [inline packages/camera/camera_web/example/integration_test/camera_test.dart:430]: Are you still tracking this down @ABausG?

Comment (ABausG) [reply packages/camera/camera_web/example/integration_test/camera_test.dart:430]: I've managed to get this test to work 🙌

File: packages/camera/camera_web/example/integration_test/camera_test.dart
Comment (ABausG) [inline packages/camera/camera_web/example/integration_test/camera_test.dart:320]: I could not get this test to work. As it does not seem to trigger the datavailable event of the media recorder.
I was able to manually test the behaviour with setting a max duration (I tested with 4 seconds) however this test seems to timeout waiting for the first element of the event stream.
Could you take a look at what's might be going on? 
Also is there a (nice) way to debug the integration tests?

File: packages/camera/camera_web/example/integration_test/camera_test.dart
Comment (bselwe) [inline packages/camera/camera_web/example/integration_test/camera_test.dart:314]:
```suggestion
        expect(camera.mediaRecorder!.state, equals('recording`));
```

File: packages/camera/camera_web/example/integration_test/camera_test.dart
Comment (bselwe) [reply packages/camera/camera_web/example/integration_test/camera_test.dart:320]: I will have a look at this test.

There is a way to debug the integration tests in the browser. You'll need to execute this in `camera_web/example`:

```
flutter run -d web-server --target integration_test/camera_test.dart --debug
```

It will produce a link that you should open in the browser, open inspection tools (CMD+Option+J in Chrome), find a dart file you want to debug in the "Source" tab and mark a breakpoint. Refresh the web page and it should stop at the breakpoint.

File: packages/camera/camera_web/example/integration_test/camera_test.dart
Comment (bselwe) [inline packages/camera/camera_web/example/integration_test/camera_test.dart:354]: nit 
```suggestion
            throwsA(predicate<PlatformException>(
                (ex) => ex.code == CameraErrorCode.notSupported.toString(),),),);
```

File: packages/camera/camera_web/example/integration_test/camera_test.dart
Comment (bselwe) [inline packages/camera/camera_web/example/integration_test/camera_test.dart:371]:
```suggestion
        expect(camera.mediaRecorder!.state, equals('paused'));
```

File: packages/camera/camera_web/example/integration_test/camera_test.dart
Comment (bselwe) [inline packages/camera/camera_web/example/integration_test/camera_test.dart:409]:
```suggestion
        expect(camera.mediaRecorder!.state, equals('recording'));
```

File: packages/camera/camera_web/example/integration_test/camera_test.dart
Comment (bselwe) [inline packages/camera/camera_web/example/integration_test/camera_test.dart:469]: nit 
```suggestion
            throwsA(predicate<PlatformException>((ex) =>
                ex.code ==
                CameraErrorCode.mediaRecordingNotStarted.toString(),),),);
      });
```

File: packages/camera/camera_web/example/integration_test/camera_web_test.dart
Comment (bselwe) [inline packages/camera/camera_web/example/integration_test/camera_web_test.dart:793]: nit
```suggestion
        expect(video, equals(capturedVideo));
```

**Comment by bselwe** [comment]
Hey @ABausG, could you give me push access to your fork so I could continue working on this?

**Comment by bselwe** [comment]
@ABausG Is there anything that you think might be missing for the video recording feature?

**Comment by ABausG** [comment]
Apart from the one test I think it should be complete

**Comment by bselwe** [comment]
@googlebot I consent.

**Comment by ditman** [comment]
I've tried to update this with the latest master to look at the test that is commented out, but I think there's other problem with the branch; I think there might be a problem regarding `camera.dispose()` (and I might have been the cause, of course :P)

`camera_test.dart` wouldn't finish, and it seems to me that the culprit were the two tests that have `await camera.dispose()`?

(not entirely sure it has to do with the tests that are failing in this PR)

**Comment by bselwe** [comment]
Yeah, thanks for checking @ditman, I'll fix that. It seems that after adding `await` to  `_videoRecorderController.close` in `Camera.dispose` tests fail. Updating `_videoRecorderController` to a broadcast stream controller fixes this issue.

**Comment by ditman** [comment]
> Updating `_videoRecorderController` to a broadcast stream controller fixes this issue.

Ahhh, very good catch there @bselwe!

File: packages/camera/camera_web/lib/src/camera.dart
Comment (ditman) [inline packages/camera/camera_web/lib/src/camera.dart:452]: This should also check if we have recorded > time than maxVideoDuration, right?

**Review by ditman — CHANGES_REQUESTED** [review]
I have 2 general comments about this PR; I don't understand very well how these features are currently implemented:

* How does this respect the max length of a video recording passed by a user? I don't think we ever check or use the value of max length (only null checks).
* How does this combine several `ondataavailable` events with video "chunks" to produce one single, large video file that can be used later? It seems this code emits the File as soon as the first `ondataavailable` event fires.
  * See the how they use `ondataavailable` and `stop` events [here](https://developer.mozilla.org/en-US/docs/Web/API/MediaRecorder/ondataavailable#example) (it's for audio, but the API is the same).
  * Is this even supported by XFile, do we need to change anything there? Do we need to create a File from an html.File, or a bunch of Blobs or something other than "bytes"?
* @bselwe noticed that video on Chrome/Edge is missing some metadata, and the duration is listed as `Infinity`; have you noticed this? It seems to be [this bug](https://bugs.chromium.org/p/chromium/issues/detail?id=642012). [SO1](https://stackoverflow.com/questions/38443084/how-can-i-add-predefined-length-to-audio-recorded-from-mediarecorder-in-chrome/39971175#39971175), [SO2](https://stackoverflow.com/questions/50586612/seeking-is-not-working-in-recorded-video-with-mediarecorder-api). (Is this something that can be fixed when putting together the XFile from all the "chunks"? Does the mp4 codec have this problem?)

For now, I'm going to publish the camera_web plugin as is, and when this PR matures a little bit more, we can release it in a new version. This is a great start, but I'm a little concerned about the comments above :S

**Comment by ditman** [comment]
> How does this respect the max length of a video recording passed by a user? I don't think we ever check or use the value of max length (only null checks).

We discussed this, and it seems that we're leveraging the `timeslice` feature (that I had misunderstood), and we only use one blob of video, until we call stop, or the `timeslice` is done.

> How does this combine several ondataavailable events with video "chunks" to produce one single, large video file

This is already achieved by the behavior of the `ondataavailable` event, and @bselwe just made it so that we use the stop event to collect the recorded video, closer to the example.

**Comment by ditman** [comment]
@googlebot I consent.

**Review by ditman — APPROVED** [review]
LGTM! All tests pass locally, let's go with this one!

**Pull request merged**
